warehouse/query-core/src/main/java/datawave/query/util/sortedmap/BufferedFileBackedSortedMap.java

```diff
@@ -0,0 +1,560 @@
+                            hdfsHandlerFactory.mapFileCount(hdfsHandlerFactory.getFileCount() - 1);
+                        }
+                    }
+                } else {
+                    log.warn("Unable to create a file handler using the handler factory: " + handlerFactory);
+                }
+            }
+
+            // if the buffer was not persisted, throw an exception
+            if (!buffer.isPersisted())
+                throw new IOException("Unable to persist the sorted map using the configured handler factories.");
+
+            buffer = null;
+            compact(maxOpenFiles);
+        }
+    }
+
+    protected List<FileSortedMap<K,V>> getMaps() {
+        List<FileSortedMap<K,V>> maps = new ArrayList<>();
+        for (SortedMap<K,V> submap : map.getMaps()) {
+            maps.add((FileSortedMap<K,V>) submap);
+        }
+        return maps;
+    }
+
+    protected void addMap(FileSortedMap<K,V> submap) {
+        map.addMap(submap);
+        size += submap.size();
+    }
+
+    public boolean hasPersistedData() {
+        for (SortedMap<K,V> submap : map.getMaps()) {
+            if (((FileSortedMap<K,V>) submap).isPersisted()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public boolean isPersisted() {
+        // we are (completely) persisted iff the buffer is persisted
+        return (buffer == null || buffer.isPersisted());
+    }
+
+    @Override
+    public int size() {
+        if (sizeModified) {
+            this.size = map.size();
+            sizeModified = false;
+        }
+        return this.size;
+    }
+
+    public int getBufferPersistThreshold() {
+        return this.bufferPersistThreshold;
+    }
+
+    public int getBufferSize() {
+        return (this.buffer == null ? 0 : this.buffer.size());
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return size() == 0;
+    }
+
+    @Override
+    public boolean containsKey(Object o) {
+        // try the cheap operation first
+        if (buffer != null && buffer.containsKey(o)) {
+            return true;
+        } else {
+            return map.containsKey(o);
+        }
+    }
+
+    @Override
+    public boolean containsValue(Object value) {
+        return false;
+    }
+
+    private String printHandlerFactories() {
+        return String.join(", ", handlerFactories.stream().map(SortedMapFileHandlerFactory::toString).collect(Collectors.toList()));
+    }
+
+    /**
+     * If the number of maps is over maxFiles, then start compacting those files down. The goal is to get the number of files down around 50% of maxFiles.
+     *
+     * @param maxFiles
+     *            the max number of files
+     * @throws IOException
+     *             for IO Exceptions
+     */
+    public void compact(int maxFiles) throws IOException {
```

Comment:
I think this does what it say it does. What I'm not clear on is why its necessary. These files should be relatively short lived. What problem are we avoiding by making multiple passes over the data to compact them?

Reply:
These files are NOT short lived.  We need a compaction mechanism to avoid the problem of too many open file handles when we do the final merge sort.